### ISSUE TITLE
Problem on rotate

### ISSUE THREAD

**Issue by acquariusoft** [jfeinstein10/SlidingMenu#77]
HI

i use xml for enable slidemenu

with xml like

<?xml version="1.0" encoding="utf-8"?>
<com.slidingmenu.lib.SlidingMenu xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:sliding="http://schemas.android.com/apk/res-auto"
    android:id="@+id/slidingmenulayout"
    android:layout_width="fill_parent"
    android:layout_height="fill_parent"
    sliding:viewAbove="@layout/xxxx"
    sliding:viewBehind="@layout/xxxx"
    sliding:behindWidth="@dimen/menu_slide_width"
    />

and se android:configChanges in androidManifest.xml

if i rotate the app the slidemenu not have the right with and sometimes it not work

Tnx for your great work!!

**Comment by slidese** [comment]
I'm not sure if I am having the same problem as you are describing above but I noticed that my sample app would not handle the width of the menu correctly when I'd flip the phone to landscape and the immediately click the home icon to reveal the menu.

It seem to be related to Android not loading the resources correctly.

Removing the android:configChanges="orientation|screenSize|screenLayout" form the manifest seem to have solved it.

**Comment by cnordvik** [comment]
I have the same problem where I actually need to have the android:configChanges="orientation|screenSize|screenLayout" of the activity. Then the menu is not displayed correctly, it seems to use the size of the previous orientation. 

Is there a way to notify the menu that it needs to recalculate it's size because the device rotated?

**Comment by tobrun** [comment]
I'm also having problems with this. I have got a activity setup with androd:configChanges, this activity is responsible for managing my fragments. 

The swipe implemenation works great but I have also setup the hardware options button to show the slidingmenu 
(by invoking the toggle() function) and replaced the ActionBar icon to do the same. The problem occurs when the function toggle() is called after an orientation change. 

Is someone looking into this?

**Comment by dani-** [comment]
Joining the #77 club.
And I think, I have a fix.

I'm using sliding menu on right side of screen and I have this in the manifest for my activity:
    `android:configChanges="orientation | keyboardHidden | screenSize"`

#### Reproducing the problem.
- Reached my activity screen in portrait mode. Tapped "menu" button (I have one on screen). Menu works perfect here. 
- Closed menu and changed orientation to landscape. Tapped "menu" again. ContentAbove is only partially open, while the menu content fills the full screen width (menu is behind and only partially visible).
- Closed menu and opened again - Everything is good now.
- Now closed menu again, changed to portrait mode. Tapped "menu". Menu content is moved way to the left. UI looks messed up.
- Closed "menu" and open again - Everything is good again.

The problem is that, scroller is not getting the correct value for scrollX on orientation change.

I cannot remove `android:configChanges="orientation | keyboardHidden | screenSize"` as I need this for various other reasons.

#### Fixing

Added the below code to `public void onConfigurationChanged(Configuration newConfig)` in my activity.

``` java
int offset = (int) getResources().getDimension(R.dimen.menu_offset); // Menu offset dp
float density = getResources().getDisplayMetrics().density;
int w = (int)(newConfig.screenWidthDp * density); // New width
int h = (int)(newConfig.screenHeightDp * density); // New height
sm.getMenu().layout(0, 0, w - offset, h); // Update content for CustomViewBehind
```

`sm` is the `SlidingMenu` reference.

Actual fix should go to SlidingMenu:
**1. SlidingMenu.java**
Add this

``` java
    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        if (w != oldw){
            mViewBehind.resetSize(w, h);
        }
    }
```

**2. CustomViewBehind.java**
Add this

``` java
    public void resetSize(int width, int height) {
        mContent.layout(0, 0, width-mWidthOffset, height);
        if (mSecondaryContent != null)
            mSecondaryContent.layout(0, 0, width-mWidthOffset, height);
    }
```

Replace line# 131-133 with an invoke to the new method (`resetSize`)

jfeinstein - Can you check if this is correct and if yes, add it to SlidingMenu.

_Note: I love SlidingMenu._

**Comment by menny** [comment]
@danimonk, _screenWidthDp_ field is available in API level 13 and higher.
I'm wonderring if this will be enough (for the `public void onConfigurationChanged(Configuration newConfig)` code):

```
int offset = (int) getResources().getDimension(R.dimen.menu_content_offset);
DisplayMetrics dm = getResources().getDisplayMetrics();
int w = dm.widthPixels;
int h = dm.heightPixels;
menu.getMenu().layout(0, 0, w - offset, h); // Update content for CustomViewBehind
```

**Comment by dani-** [comment]
@menny you are correct. Need to find an option to support lower versions as well.

**Comment by rogerheim** [comment]
@danimonk lines 131-133 are in onMeasure(). Can't remove setMeasuredDimension(); what is the change you're making there?

**Comment by dani-** [comment]
@rogerheim 
Looks like the code has changed after I posting the comment. 
My suggestion was to extract the following lines of code from `onLayout` method

``` java
mContent.layout(0, 0, width-mWidthOffset, height);
if (mSecondaryContent != null)
     mSecondaryContent.layout(0, 0, width-mWidthOffset, height);
```

to the method `resetSize` (to be added new), so that this piece of code can be invoked from `onSizeChanged` (newly suggested method) in `SlidingMenu`.

:+1: _Lesson learned: Do not use line numbers while explaining code_ :)

**Comment by rogerheim** [comment]
@dani- Got it. Thanks for the reply.

**Comment by vinothbabu** [comment]
@dani- tried what you indicated but no luck. 

``` Java
public void onConfigurationChanged(Configuration newConfig){
    super.onConfigurationChanged(newConfig);
    int offset = (int) getResources().getDimension(R.dimen.slidingmenu_offset); // Menu offset dp
    float density = getResources().getDisplayMetrics().density;
    int w = (int)(newConfig.screenWidthDp * density); // New width
    int h = (int)(newConfig.screenHeightDp * density); // New height
    getSlidingMenu().getMenu().layout(0, 0, w - offset, h);
}
```

Added the above code to my Activity class. 

``` Java
@Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        if (w != oldw){
            mViewBehind.resetSize(w, h);
        }
    }
```

Added the above to `slidingmenu.java`

``` Java
public void resetSize(int width, int height) {
        mContent.layout(0, 0, width-mWidthOffset, height);
        if (mSecondaryContent != null)
            mSecondaryContent.layout(0, 0, width-mWidthOffset, height);
}
```

Added the above to `CustomViewBehind.java`

**Comment by Ultimater** [comment]
@dani- when we invoke your `resetSize()` method, what parameters should we be sending it? `0, 0`, `width,height` or `width-mWidthOfset,height`? I'm guessing you want us to edit [**CustomViewBehind.java**](https://github.com/jfeinstein10/SlidingMenu/blob/6ca0c5e1e5067a29aa64408bb2d6d62fe0a726bb/library/src/com/slidingmenu/lib/CustomViewBehind.java) and find and replace

``` Java
mContent.layout(0, 0, width-mWidthOffset, height);
if (mSecondaryContent != null)
     mSecondaryContent.layout(0, 0, width-mWidthOffset, height);
```

and edit it with:

``` Java
@Override
 protected void onLayout(boolean changed, int l, int t, int r, int b) {
  final int width = r - l;
  final int height = b - t;
   mContent.resetSize(width, height);
 }
```

Is that correct?

**Comment by Nandan078** [comment]
Dani, Jeremy - can any of you please help on this one if you have a solution?

**Comment by dani-** [comment]
@vinothbabu 
The changes to `onConfigurationChanged` method is a work around.
You may choose one of the option below - not both at same time.

**Option A**
Add/Modify `onConfigurationChanged` method as described above.

**Option B**
1. Add `onSizeChanged` method to SlidingMenu.java
2. Add `resetSize` method to CustomViewBehind.java

**Comment by dani-** [comment]
@Ultimater 
You are correct, except that in `onLayout` method, it is not

``` java
mContent.resetSize(width, height);
```

but just

``` java
resetSize(width, height);
```

**Comment by dani-** [comment]
@Nandan078 were you referring to questions by @vinothbabu and @Ultimater ?

**Comment by Nandan078** [comment]
Yes Dani - was referring to the same things as Vinoth and Ultimater, thank you for the response. Will try your suggestion and keep you posted...
